### PR TITLE
Implemented peak mapping handling into RDC restraint/transformer

### PR DESCRIPTION
**Pull request by splenz**
Trying to test how well rdc restraints work with peak mapping objects:

Ported _handle_mapping from meld/transform/restraints/meld/meld.py to meld/transform/restraints/rdc.py; added peak mapping logic to RDCRestraint in med/system/restraints.py

### CODE REVIEW

**Comment by jlmaccal** [comment]
I think the main challenge is how we handle cases where we either have:

- more peaks than AtomGroups
- more atom groups than peaks

There is special handling for these in the cuda code for the main meld forces. I think we have to think about how to handle this in the CustomBondForce that implements RDCs.

We probably need to add an extra flag `dummy` for each bond. If either of the atoms involved maps to -1, then we set the dummy flag and the energy is always zero. Or something like that.

**Comment by jlmaccal** [comment]
There's also a couple of typing errors, but we can fix those later. (You can run `mypy meld` from the base directory to run the type checks.)

**Comment by jlmaccal** [comment]
So, I think the case where either atom index maps to -1 needs special handling.

We need to:
- add a custom parameter called `dummy` or `skip` or something like that
- If both indices are `!= -1`, then we set `dummy = 0`
- If either index `==-1` then we set `dummy = 1`
    - In this case, we still need to set the indices to something valid. We can probably use something like the first two atoms, or one of the other atom groups.
- Inside of the energy function, we need to use a `select(dummy, 0, energy)` or something similar to give zero energy whenever `dummy == 1`.
-

File: meld/runner/transform/restraints/rdc.py [hunk 1]
```diff
@@ -87,9 +88,18 @@ def add_interactions(
                 rests = self.alignment_map[alignment]
                 force = self.alignment_forces[alignment]
                 for r in rests:
+                    i, j = self._handle_mapping([r.atom_index_1, r.atom_index_2], state)
+                    if i != -1 and j != -1:
+                        dummy = 0
+                    # If Peak is unassigned, need to set dummy to True and give the energy function atom indices that ultimately do not matter
+                    elif i == -1 or j == -1:
```
Comment (jlmaccal) [inline meld/runner/transform/restraints/rdc.py:95]: This condition is redundant, you can replace the `elif` with just `else`.

File: meld/runner/transform/restraints/rdc.py [hunk 2]
```diff
@@ -87,9 +88,18 @@ def add_interactions(
...
+                        dummy = 1
+                        i, j = 0, 1
```
Comment (jlmaccal) [inline meld/runner/transform/restraints/rdc.py:97]: Add a comment here explaining why we're doing this.

File: meld/runner/transform/restraints/rdc.py [hunk 3]
```diff
@@ -118,12 +128,20 @@ def update(
                 force = self.alignment_forces[alignment]
                 for index, r in enumerate(rests):
                     scale = r.scaler(alpha) * r.ramp(timestep)
-                    atoms, params = force.getBondParameters(index)
-                    assert atoms[0] == r.atom_index_1
+                    i, j = self._handle_mapping([r.atom_index_1, r.atom_index_2], state)
+                    if i != -1 and j != -1:
+                        dummy = 0
+                    # If Peak is unassigned, need to set dummy to True and give the energy function atom indices that ultimately do not matter
+                    elif i == -1 or j == -1:
```
Comment (jlmaccal) [inline meld/runner/transform/restraints/rdc.py:135]: Same as above, just change to `else`

File: meld/runner/transform/restraints/rdc.py [hunk 4]
```diff
@@ -118,12 +128,20 @@ def update(
...
+                        dummy = 1
+                        i, j = 0, 1
```
Comment (jlmaccal) [inline meld/runner/transform/restraints/rdc.py:137]: Add a comment.

**Review by jlmaccal — CHANGES_REQUESTED** [review]

File: meld/runner/transform/restraints/rdc.py (hunk 2 above)
Comment (jlmaccal) [reply meld/runner/transform/restraints/rdc.py:97]: This is for line 97, just explain why we're sticking 0, 1 in as the indices.

File: meld/runner/transform/restraints/rdc.py [hunk 5]
```diff
@@ -129,12 +132,15 @@ def update(
...
-                    # If Peak is unassigned, need to set dummy to True and give the energy function atom indices that ultimately do not matter
-                    elif i == -1 or j == -1:
+                    # If Peak is unassigned, need to set dummy to 1 (True)
+                    else:
+                        # If dummy is 1 then energy will always be zero 
                         dummy = 1
-                        i, j = 0, 1
+                        # These indices are arbitrary and do not matter but need something to do energy calculation
+                        i, j = 0, 1 
                     #assert atoms[0] == r.atom_index_1
```
Comment (jlmaccal) [inline meld/runner/transform/restraints/rdc.py:144]: Do we need the commented-out assertion?

Otherwise, looks good.

Comment (splenz) [reply meld/runner/transform/restraints/rdc.py]: I don't think so? I removed it in the most recent commit

**Review by jlmaccal — APPROVED** [review]

**Pull request merged**